### PR TITLE
ask_async

### PR DESCRIPTION
**Pull request by sunildkumar**
This enables customers to query `groundlight` asynchronously. It adds `want_async` as a new parameter on `submit_image_query` and  implements `ask_async`, a convenient interface to submit queries asynchronously.

Blockers to merging this:
~~1. Merge my corresponding server-side PR the enables async IQ submissions.~~
~~2. Make changes based on updates to the BE~~
3. Fix PR comments 
4. Show that new SDK tests pass after merge, right now they fail because the server side code to support this feature isn't deployed. 
5. How do we want to document this added functionality? I will have to add to the docs reference in sphinx once @blaise-muhirwa sets that up :). I should definitely show a small example here about how you need to query for results after querying async.
6. Do we want to add a section to `docs/building-applications` with a small demo of how one can use `ask_async`?

### CODE REVIEW

File: src/groundlight/client.py
Comment (sunildkumar) [inline src/groundlight/client.py:1]: In addition to this code, let me know what you think about this following:

1. How do we want to document this added functionality? I will have to add to the docs reference in sphinx once Blaise merges this. I should definitely show a small example here about how you need to query for results after querying async. 
2. Do we want to add a section to docs/building-applications with a small demo of how one can use ask_async?

File: spec/public-api.yaml
Comment (robotrapta) [inline spec/public-api.yaml:1]: Good question.

File: src/groundlight/client.py
Comment (robotrapta) [reply src/groundlight/client.py:1]: I think we should definitely sprinkle it into the top-level docs.  I’d describe an application where you have one thread uploading images and another thread processing the results.

File: src/groundlight/client.py
Comment (robotrapta) [inline src/groundlight/client.py:82]: Nice and clean.

File: src/groundlight/client.py
Comment (robotrapta) [inline src/groundlight/client.py:232]: I think here we should be more deliberate about what we send to the API.  We can be pretty flexible in what the SDK method uses to indicate that they want async - they can give us an actual bool, or a 0/1 int or something else which indicates truthiness.  But we should be more specific about what we send to the server.  In the other PR I argue for always sending “1” or “0”.

File: src/groundlight/client.py
Comment (robotrapta) [inline src/groundlight/client.py:262]: We should flesh this out with full documentation of all the parameters accepted.

I have a bunch of thoughts of how to word the summary here, but I’d rather type them when I’m at a proper keyboard.

**Review by robotrapta — APPROVED** [review]
This LGTM - some suggestions within, and of course we should be conscious of dependencies when merging.

Also, this probably deserves a major version bump to 0.12, especially with the other `ask_` methods.

File: .vscode/settings.json
Comment (blaise-muhirwa) [inline .vscode/settings.json:6]: Looks like you accidentally committed this!

Comment (sunildkumar) [reply .vscode/settings.json:6]: this was intentional - I was just adding the stuff we have in zuuul that we needed here to
1. make it easier to keep code linted
2. make black behave rationally

File: spec/public-api.yaml
Comment (sunildkumar) [reply spec/public-api.yaml:1]: Talked to mike about this - we discussed the aspirational goal of organizational/behavioral change to ensure this stays up to date.

File: src/groundlight/client.py
Comment (sunildkumar) [reply src/groundlight/client.py:82]: with the serverside change (where the serve returns result=None, this is no longer necessary, so it's even cleaner!

File: src/groundlight/client.py
Comment (sunildkumar) [reply src/groundlight/client.py:232]: Why not just True/False?

File: src/groundlight/client.py
Comment (sunildkumar) [reply src/groundlight/client.py:262]: I copy pasted the relevant descriptions from submit_image_query for now.

File: .vscode/settings.json
Comment (sunildkumar) [reply .vscode/settings.json:6]: ...at least on my machine. the rulers are helpful to make the linter happy and black does wild things with the imports if it is allowed to touch them.

File: src/groundlight/client.py
Comment (brandon-groundlight) [reply src/groundlight/client.py:350]: I'd defer to Leo. My understanding is the the "1" "0" distinction fixes the bits representation. True values can have multiple representations (that occasionally are different by system) so we might as well choose a 1:1 mapping on the values sent into our system rather than a many to one mapping. 
Again, I defer to Leo

**Review by brandon-groundlight — APPROVED** [review]

File: src/groundlight/client.py
Comment (sunildkumar) [reply src/groundlight/client.py:350]: I disagree with this pretty strongly as we are going to essentially going to have to write a two translation layers (on both the SDK and BE) to convert to and from these "string bools". OTOH, python's `True` and `False` are already unambiguous values for T/F. Additionally the SDK automatically handles serialization of `True`/`False` and our BE automatically handles deserialization to `True`/`False` so no custom translation is required.

Comment (sunildkumar) [reply src/groundlight/client.py:350]: If we really want this, I think we should do it as its own PR as that way we can focus on applying this update uniformly as there is BE work required as well.

Comment (sunildkumar) [reply src/groundlight/client.py:350]: If we want to accept other arguments to represent truthiness from the user I think the solution is to do validation on the inputs prior to whatever business logic we do.

File: src/groundlight/client.py
Comment (blaise-muhirwa) [inline src/groundlight/client.py:343]:
```suggestion
        if want_async:
```

**Review by blaise-muhirwa — APPROVED** [review]
Awesome! As far as documenting this with sphinx, it won't be a problem since sphinx autogenerates docs for everything under `client.py`. Just to make sure things look good tho you can run `make develop-docs-comprehensive` to make sure the docs look good!

File: src/groundlight/client.py
Comment (robotrapta) [reply src/groundlight/client.py:350]:
```suggestion
            params["want_async"] = str(bool(want_async))
```
is what I mean.  Don't rely on the variable's python-truthiness and its string representation being the same.

**Pull request merged**
